File: gunhoflash/lowest-common-ancestor-of-a-binary-search-tree.py
class TreeNode:
  def __init__(self, x):
    self.val = x
    self.left = None
    self.right = None

class Solution:
  def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
    def getLowestCommonAncestor(root: TreeNode, p: int, q: int) -> TreeNode:
      if p <= root.val <= q:
        return root
      if q < root.val:
        return getLowestCommonAncestor(root.left, p, q)
      return getLowestCommonAncestor(root.right, p, q)

    return getLowestCommonAncestor(root, min(p.val, q.val), max(p.val, q.val))

File: gunhoflash/lowest-common-ancestor-of-a-binary-search-tree.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function lowestCommonAncestor(
  root: TreeNode,
  p: TreeNode,
  q: TreeNode
): TreeNode {
  function getLowestCommonAncestor(
    root: TreeNode,
    p: number,
    q: number
  ): TreeNode {
    if (p <= root.val && root.val <= q) return root;
    return q < root.val
      ? getLowestCommonAncestor(root.left!, p, q)
      : getLowestCommonAncestor(root.right!, p, q);
  }

  return getLowestCommonAncestor(
    root,
    Math.min(p.val, q.val),
    Math.max(p.val, q.val)
  );
}
